Update and extend doc/RPM.md

cracksalad opened the pull request:
I just noticed some minor mismatches and an outdated dependency list in the doc/RPM.md file. So I fixed/updated that and I appended a Docker file as a simple and clean way to create a build environment for the RPM build.

I tested the updated dependency list (in a sense that all necessary packages are included) as well as the Docker file. 
There is just one thing to note: The `prometheus.sh` file can not be found during the build (file is referenced in `pgexporter.spec`), which I have pointed out in https://github.com/pgexporter/pgexporter/issues/329#issuecomment-3302494291. Not sure if that is caused by another flaw in the docs or something else though.

resyfer commented:
@cracksalad Regarding the issue, this might help:

```
[root@6722228c248d ~]# cat /var/tmp/rpm-tmp.RTl3ra | grep prometheus.sh
/usr/bin/install -m 644 /root/rpmbuild/BUILD/pgexporter-0.8.0/contrib/prometheus_scrape/prometheus.sh /root/rpmbuild/BUILDROOT/pgexporter-0.8.0-1.el9.x86_64/usr/share/doc/pgexporter/prometheus_scrape/prometheus.sh

[root@6722228c248d ~]# cat pgexporter/pgexporter.spec | grep prometheus.sh
%{__install} -m 644 %{_builddir}/%{name}-%{version}/contrib/prometheus_scrape/prometheus.sh %{buildroot}%{_docdir}/%{name}/prometheus_scrape/prometheus.sh
%{_docdir}/%{name}/prometheus_scrape/prometheus.sh

[root@6722228c248d ~]# ls /root/rpmbuild/BUILD/pgexporter-0.8.0/contrib/prometheus_scrape
README.md  extra.info  prometheus.py

[root@6722228c248d ~]# cat /var/tmp/rpm-tmp.RTl3ra | grep prometheus_scrape
/usr/bin/mkdir -p /root/rpmbuild/BUILDROOT/pgexporter-0.8.0-1.el9.x86_64/usr/share/doc/pgexporter/prometheus_scrape
/usr/bin/install -m 644 /root/rpmbuild/BUILD/pgexporter-0.8.0/contrib/prometheus_scrape/extra.info /root/rpmbuild/BUILDROOT/pgexporter-0.8.0-1.el9.x86_64/usr/share/doc/pgexporter/prometheus_scrape/extra.info
/usr/bin/install -m 644 /root/rpmbuild/BUILD/pgexporter-0.8.0/contrib/prometheus_scrape/prometheus.sh /root/rpmbuild/BUILDROOT/pgexporter-0.8.0-1.el9.x86_64/usr/share/doc/pgexporter/prometheus_scrape/prometheus.sh
/usr/bin/install -m 644 /root/rpmbuild/BUILD/pgexporter-0.8.0/contrib/prometheus_scrape/README.md /root/rpmbuild/BUILDROOT/pgexporter-0.8.0-1.el9.x86_64/usr/share/doc/pgexporter/prometheus_scrape/README.md

[root@6722228c248d ~]# ls pgexporter/contrib/prometheus_scrape/
README.md  extra.info  prometheus.py
```

This is probably because `prometheus.sh` was changed to `prometheus.py` and this was overlooked in the package. Thanks for this issue, this would have gone unnoticed if not for this.

cracksalad commented:
If it is just a replacement of `.sh` to `.py` I could totally add that to this PR or another one, if you prefer that. Should I go ahead?

resyfer commented:
@cracksalad There is a change that I feel is required. You remove `coreutils-single` due to conflicts with `coreutils` planning to install it later, but commands following it fail due to coreutils missing any way :) .

I get something like:

```
65.34 error: Error -1 running transaction
------
Dockerfile:5
--------------------
   4 |
   5 | >>> RUN rpm -e --nodeps coreutils-single && \
   6 | >>>   microdnf upgrade -y && \
   7 | >>>   microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint coreutils diffutils patch rpmdevtools chrpath && \
   8 | >>>   microdnf install --enablerepo=crb -y zlib-devel bzip2-devel lz4-devel libev-devel openssl-devel python3-docutils doxygen libyaml-devel lz4 systemd-devel && \
   9 | >>>   cd ~ && rpmdev-setuptree && \
  10 | >>>   git clone https://github.com/cracksalad/pgexporter.git && \
  11 | >>>   echo $'(mkdir ~/pgexporter/build && \
  12 | >>>   cd ~/pgexporter/build && \
  13 | >>>   cmake -DCMAKE_BUILD_TYPE=Release .. && \
  14 | >>>   make package_source && \
  15 | >>>   VERSION=$(grep -Po "Version:\s*\K(\d+\.\d+\.\d+)" ~/pgexporter/pgexporter.spec) && \
  16 | >>>   cp pgexporter-$VERSION.tar.gz ~/rpmbuild/SOURCES/$VERSION.tar.gz && \
  17 | >>>   cd ~/pgexporter && \
  18 | >>>   QA_RPATHS=0x0001 rpmbuild -bb pgexporter.spec) || \
  19 | >>>   cd ~' > build.sh && chmod +x build.sh
  20 |
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c rpm -e --nodeps coreutils-single &&   microdnf upgrade -y &&   microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint coreutils diffutils patch rpmdevtools chrpath &&   microdnf install --enablerepo=crb -y zlib-devel bzip2-devel lz4-devel libev-devel openssl-devel python3-docutils doxygen libyaml-devel lz4 systemd-devel &&   cd ~ && rpmdev-setuptree &&   git clone https://github.com/cracksalad/pgexporter.git &&   echo $'(mkdir ~/pgexporter/build &&   cd ~/pgexporter/build &&   cmake -DCMAKE_BUILD_TYPE=Release .. &&   make package_source &&   VERSION=$(grep -Po \"Version:\\s*\\K(\\d+\\.\\d+\\.\\d+)\" ~/pgexporter/pgexporter.spec) &&   cp pgexporter-$VERSION.tar.gz ~/rpmbuild/SOURCES/$VERSION.tar.gz &&   cd ~/pgexporter &&   QA_RPATHS=0x0001 rpmbuild -bb pgexporter.spec) ||   cd ~' > build.sh && chmod +x build.sh" did not complete successfully: exit code: 1
```

I tried a lot of various ways, but I couldn't get it to work with microdnf. There's a flag `--allowerasing` that can swap out conflicting packages atomically, but that's not available in alma linux's microdnf (haven't checked others but I'm positive microdnf allows it). Can you please take a look at it?

Also, please split the RUN command into separate RUN commands for install and clone-and-build to allow quicker building due to docker layer caching. Optionally, you can keep the image minimal by removing dnf cache after install (I think by `dnf clean all && rm -rf /var/cache/yum`).

resyfer commented:
@cracksalad I tried separating things into separate commands:

```
...
 => ERROR [3/8] RUN microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint diffutils patch rpmdevtools chrpath                                                            34.0s
------
 > [3/8] RUN microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint diffutils patch rpmdevtools chrpath:
0.755 Package                                                  Repository     Size
0.755 Installing:
0.755  abattis-cantarell-fonts-0.301-4.el9.noarch              appstream  372.8 kB
0.755  acl-2.3.1-4.el9.x86_64                                  baseos      70.4 kB
...
1.598 Downloading packages...
24.48 Running transaction test...
26.96 Installing: elfutils-libelf;0.192-6.el9_6.alma.1;x86_64;baseos
27.42 Installing: findutils;1:4.8.0-7.el9;x86_64;baseos
...
33.76 Installing: chrpath;0.16-17.el9;x86_64;baseos
33.78 error: Error -1 running transaction
------
Dockerfile:6
--------------------
   4 |
   5 |     RUN microdnf upgrade -y
   6 | >>> RUN microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint diffutils patch rpmdevtools chrpath
   7 |     RUN microdnf install --enablerepo=crb -y check-devel zlib-devel bzip2-devel lz4-devel libev-devel openssl-devel python3-docutils doxygen libyaml-devel lz4 systemd-devel
   8 |     RUN microdnf clean all && rm -rf /var/cache/yum && \
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c microdnf install -y git gcc cmake make rpm-build rpm-devel rpmlint diffutils patch rpmdevtools chrpath" did not complete successfully: exit code: 1
```

Hmmm, this sort of error is apparently from a kernel too old (of host?) @jesperpedersen @bassamadnan can you guys confirm the docker file please?

I also reordered some things:

```
...
RUN cd ~ && git clone https://github.com/cracksalad/pgexporter.git && cd pgexporter/contrib/docker/rpm && cp ./build.sh /root/build.sh

RUN chmod +x /root/build.sh
...
```

cracksalad commented:
> I also reordered some things:
> 
> ```
> ...
> RUN cd ~ && git clone https://github.com/cracksalad/pgexporter.git && cd pgexporter/contrib/docker/rpm && cp ./build.sh /root/build.sh
> 
> RUN chmod +x /root/build.sh
> ...
> ```

Do you want to replace the `COPY` command with taking the `build.sh` from the cloned repo or was that just for testing?

___

I split (basically) all the commands into their own `RUN` statement and added the `:Z` to the doc. 

___

During validation of these changes, I came across the following CMake error with the current main branch:

```
-- Performing Test HAS_RELRO - Success
-- Performing Test HAS_NOW
-- Performing Test HAS_NOW - Success
CMake Error at CMakeLists.txt:203 (add_subdirectory):
  add_subdirectory given source "test" which is not an existing directory.


-- Configuring incomplete, errors occurred!
```

Although, I was able to validate with the `0.7.0` tag. So this is probably out of the scope of this PR, just want to let you know.

resyfer commented:
@cracksalad Ahhh ok, my bad, I had not considered doing it in the repository itself. Alright, that's fine by me, makes more sense than what I was trying. I will merge it in a bit.
